### PR TITLE
Fixed Stored XSS on Hichat

### PR DESCRIPTION
**Pull request by d3m0n-r00t**
### ⚙️ Description *
The ```HiChat``` application is a simple ```chat``` for ```small teams``` which is based on a ```small server``` which helps to ```track``` messages sent by various users. The project presents a ```stored XSS``` issue which could lead to ```session stealing``` or ```HTML/JS injection``` in a permanent way, until the victim leaves the chat room.
Fixed the stored xss using the sanitize-html module to filter out html tags from the input. 
https://www.npmjs.com/package/sanitize-html

### 💻 Technical Description *
Sanitize-html is a simple sanitizer module, to filter out html tags and sanitize input.

### 🐛 Proof of Concept (PoC) *
  1. Download the project locally
  2. Run ```npm i```
  3.  Run ```node server```
  4. Go on ```http://localhost:3000```
  5.  Join the chat room and set your username as ```test">```
  6.  Open an anonymous tab and join the chatroom as another user (simulating the victim who joins the same chat room of the attacker)
  7.  When the attacker sends any message, the username is insecurely reflected and JS injected
  8. ``` XSS popped !!!```
hichatpoc

### 🔥 Proof of Fix (PoF) *
Added sanitize-html to sanitize the input and filter out html tags to avoid injecting xss payloads into the ```nickname``` field as well as the ```message``` field. There is no xss after sanitization. 
hichatpof

### 👍 User Acceptance Testing (UAT)
The application runs smoothly after the fix, and no more xss. Can now send and receive messages without worrying about ```xss``` or ```data leakage```.

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
LGTM

**Review by Mik317 — APPROVED** [review]
LGTM :smile: :tada:

Cheers,
Mik

**Comment by alromh87** [comment]
Unfortunately this fix is bypassable using socket io directly

POC:
1. Run server
2. Open chat from browser
3. create client.js
```
const
    io = require("socket.io-client"),
    ioClient = io.connect("http://localhost:3000/");

ioClient.on("system", (usr,_2,_3) => console.info('system: ', usr));
ioClient.on("newMsg", (usr,msg) => console.info(usr, msg));

ioClient.on('connect', () => {
   console.log('Conectado')
   ioClient.emit('login', 'testUser"><img/src="x"/onerror="alert(\'Login: Hello from console\')">')
   ioClient.emit('postMsg', 'testMesage"><img/src="x"/onerror="alert(\'Msg: Hello from console\')">')
   ioClient.emit('postMsg', '<script>alert("Msg script")</script>')
});
``` 
4. Execute js
```
node client.js
```
5. Script is exected on web browser client

Captura de pantalla de 2020-09-04 14-58-20

**Comment by ghost** [comment]
Ah interesting - didn't think to try that, cheers. @d3m0n-r00t thoughts?

**Comment by Mik317** [comment]
@alromh87 you're right :smile:
I bypassed a similar fix an hour ago with the same technique :joy: 
Validation `server side` always wins :+1: 

Cheers,
Mik

**Comment by alromh87** [comment]
Sent updated PR using library

Relevant lines is where data is displayed to user:

https://github.com/418sec/HiChat/commit/c085762871d858ceb18df75b22ca48c2c117552f#diff-a90151b03560b4909234dfd89c452458R146

**Comment by d3m0n-r00t** [comment]
@alromh87, 
Thanks for letting me know about the bypass. I hadn't though about this attack mode. 
I have added sanitization in input in the server side too. Now it sanitizes input from browser as well as console. 
@Mik317 , @418raj . Please let me know if any problem with the fix. 
Proof-of-fix
bypasspof

PS: It is a wonderful bypass. It took me some time reproducing the bug.

File: server.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ var express = require('express'),
     app = express(),
     server = require('http').createServer(app),
     io = require('socket.io').listen(server),
+    sanitizehtml = require('./www/scripts/sanitizehtml.js')
```
Comment (Mik317) [inline server.js:5]: I'm still able to inject some `HTML tags` which could lead to `deface` :smile:

Screenshot from 2020-09-04 17-52-30

Also, the `emoji` functionality doesn't work well :cry: 

Cheers,
Mik

**Review by Mik317 — CHANGES_REQUESTED** [review]
Just rechecked:
* HTML tags are still injectable in part (used the payloads on https://cure53.de/purify) even if no XSS possible
* emoji functionality broken

Cheers,
Mik

**Comment by d3m0n-r00t** [comment]
@Mik317, Nice find. I accidentally sanitized out the emoji. I have added the fix. Cheers.
About you were able to inject ```html``` tags. I have added sanitization in every input. So all the tags that might cause a problem is filtered out. The ```deface``` you say is just the residual gibberish that comes after sanitization and is limited to the message box not the entire website. It is just like a user typing in unwanted messages. It doesn't affect users. I believe that it is not enough to cause a problem and the app seems to be working fine even after the injection. 
Please let me know if I am wrong. 
Thanks

**Review by Mik317 — APPROVED** [review]
LGTM :smile: :tada:

Cheers,
Mik

**Comment by huntr-helper** [comment]
Congratulations d3m0n-r00t - your fix has been selected! 🎉

Thanks for being part of the community & helping secure the world's open source code.
If you have any questions, please respond in the comments section. Your bounty is on its way - keep hunting!

**Pull request merged**